lib/ansible/modules/cloud/amazon/elb_application_lb.py

```diff
@@ -0,0 +1,715 @@
+                                if listener['Rules']:
+                                    if current_rule not in listener['Rules']:
+                                        rules_to_remove.append(current_rule)
+                                else:
+                                    rules_to_remove.append(current_rule)
+
+                            # for rules to remove we need to lookup the rule arn using unique attributes.
+                            # I have used path and priority
+                            if rules_to_remove:
+                                rule_arns_to_remove = []
+                                current_rules = connection.describe_rules(ListenerArn=current_listener['ListenerArn'])['Rules']
+                                # listener_changed = True
+                                for rules in rules_to_remove:
+                                    for current_rule in current_rules:
+                                        # if current_rule['Priority'] != 'default':
+                                        if current_rule['Conditions'] == rules['Conditions'] and current_rule['Priority'] == rules['Priority']:
+                                            rule_arns_to_remove.append(current_rule['RuleArn'])
+
+                                listener_changed = True
+                                for arn in rule_arns_to_remove:
+                                    connection.delete_rule(RuleArn=arn)
+
+                            rules_to_add = []
+                            if listener['Rules']:
+                                for rules in listener['Rules']:
+                                    if rules not in current_rules_array:
+                                        rules_to_add.append(rules)
+
+                            if rules_to_add:
+                                listener_changed = True
+                                for rule in rules_to_add:
+                                    rule['ListenerArn'] = current_listener['ListenerArn']
+                                    rule['Priority'] = int(rule['Priority'])
+                                    for action in rule['Actions']:
+                                        action['TargetGroupArn'] = convert_tg_name_arn(connection, module, action['TargetGroupName'])
+                                        del action['TargetGroupName']  
+                                    connection.create_rule(**rule)
+
+        else:
+            module.exit_json(list=listeners)
+            for listener in listeners:
+                listener['LoadBalancerArn'] = elb['LoadBalancerArn']
+                if 'Rules' in listener.keys():
+                    del listener['Rules']
+
+                # handle default
+                for default_action in listener['DefaultActions']:
+                    default_action['TargetGroupArn'] = convert_tg_name_arn(connection, module, default_action['TargetGroupName'])  
```

Comment:
@naslanidis can you explain what's going on here?  I'm unsure what the convert_tg_name_arn is for - the ARN is part of listener.  When I run the module i get the following failure here:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'TargetGroupName'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 713, in <module>\n    main()\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 708, in main\n    create_or_update_elb(connection, connection_ec2, module)\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 621, in create_or_update_elb\n    listener_changed = create_or_update_elb_listeners(connection, module, elb)\n  File \"/tmp/ansible_YZ6wr2/ansible_module_elb_application_lb.py\", line 502, in create_or_update_elb_listeners\n    default_action['TargetGroupArn'] = convert_tg_name_arn(connection, module, default_action['TargetGroupName'])  \nKeyError: 'TargetGroupName'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
```